File: bindings/fsharp/bench/EveraniumItb.FSharp.Bench/BenchStream.fs
// Stream-pump throughput vs plaintext size (Streaming Non-AEAD
// profile) at 1 MiB / 16 MiB / 64 MiB.

module EveraniumItb.FSharp.Bench.BenchStream

open System.IO
open EveraniumItb.FSharp

let run () : unit =
    use pipe =
        ItbError.get (Pipeline.init (BenchUtil.profileName "streaming-noaead-triple-v1") (BenchUtil.buildOpts ()))

    BenchUtil.header ()

    for size in BenchUtil.sizes do
        let plain = BenchUtil.payload size

        BenchUtil.benchCase "stream_pump" size (fun () ->
            use wire = new MemoryStream(size + size / 4 + 131_072)
            ItbError.get (Stream.pumpEncrypt pipe (new MemoryStream(plain, false)) wire))

File: bindings/scala/bench/src/main/scala/dev/everanium/itb/bench/BenchStream.scala
// Stream-pump throughput vs plaintext size (Streaming Non-AEAD
// profile) at 1 MiB / 16 MiB / 64 MiB.

package dev.everanium.itb.bench

import java.io.{ByteArrayInputStream, ByteArrayOutputStream}

import scala.util.Using

import dev.everanium.itb.Pipeline

object BenchStream:

  def run(): Unit =
    val profile = BenchUtil.profileName("streaming-noaead-triple-v1")
    Using.resource(
      Pipeline.init(profile, BenchUtil.buildOpts).fold(e => throw e, identity)
    ) { pipe =>
      BenchUtil.header()
      for size <- BenchUtil.Sizes do
        val plain = BenchUtil.payload(size)
        BenchUtil.benchCase("stream_pump", size) {
          val wire = new ByteArrayOutputStream(size + size / 4 + 131_072)
          pipe
            .encryptStreamPump(new ByteArrayInputStream(plain), wire)
            .fold(e => throw e, _ => ())
        }
        // Pre-encrypt one wire outside the decrypt timing loop.
        val setupWire = new ByteArrayOutputStream(size + size / 4 + 131_072)
        pipe
          .encryptStreamPump(new ByteArrayInputStream(plain), setupWire)
          .fold(e => throw e, _ => ())
        val decWire = setupWire.toByteArray
        BenchUtil.benchCase("stream_pump-dec", size) {
          val out = new ByteArrayOutputStream(size + 131_072)
          pipe
            .decryptStreamPump(new ByteArrayInputStream(decWire), out)
            .fold(e => throw e, _ => ())
        }
    }
